precompiles/dapps-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,849 @@
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! Astar dApp staking interface.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use fp_evm::PrecompileHandle;
+use parity_scale_codec::MaxEncodedLen;
+
+use frame_support::{
+    dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+    ensure,
+    traits::ConstU32,
+};
+
+use pallet_evm::AddressMapping;
+use precompile_utils::{
+    prelude::*,
+    solidity::{
+        codec::{Reader, Writer},
+        Codec,
+    },
+};
+use sp_core::{Get, H160, U256};
+use sp_runtime::traits::Zero;
+use sp_std::{marker::PhantomData, prelude::*};
+extern crate alloc;
+
+use astar_primitives::{dapp_staking::SmartContractHandle, AccountId, Balance};
+use pallet_dapp_staking_v3::{
+    AccountLedgerFor, ActiveProtocolState, ContractStake, ContractStakeAmount, CurrentEraInfo,
+    DAppInfoFor, EraInfo, EraRewardSpanFor, EraRewards, IntegratedDApps, Ledger,
+    Pallet as DAppStaking, ProtocolState, SingularStakingInfo, StakerInfo, Subperiod,
+};
+
+pub const STAKER_BYTES_LIMIT: u32 = 32;
+type GetStakerBytesLimit = ConstU32<STAKER_BYTES_LIMIT>;
+
+pub type DynamicAddress = BoundedBytes<GetStakerBytesLimit>;
+
+#[cfg(test)]
+mod test;
+
+/// Helper struct used to encode protocol state.
+#[derive(Debug, Clone, solidity::Codec)]
+pub(crate) struct PrecompileProtocolState {
+    era: U256,
+    period: U256,
+    subperiod: u8,
+}
+
+/// Helper struct used to encode different smart contract types for the v2 interface.
+#[derive(Debug, Clone, solidity::Codec)]
+pub struct SmartContractV2 {
+    contract_type: SmartContractTypes,
+    address: DynamicAddress,
+}
+
+/// Convenience type for smart contract type handling.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) enum SmartContractTypes {
+    Evm,
+    Wasm,
+}
+
+impl TryFrom<u8> for SmartContractTypes {
+    type Error = Revert;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::Evm),
+            1 => Ok(Self::Wasm),
+            _ => Err(RevertReason::custom("Unknown smart contract type").into()),
+        }
+    }
+}
+
+impl Into<u8> for SmartContractTypes {
+    fn into(self) -> u8 {
+        match self {
+            Self::Evm => 0,
+            Self::Wasm => 1,
+        }
+    }
+}
```

Comment:
Thanks, comments applied.

I tried to specify the custom error as `Revert` but it would require me to add a new struct just to handle it via the existing macro so I just added a `map_err` to the function that needs it.